src/crimson/os/seastore/onode_manager/staged-fltree/node.cc

```diff
@@ -1809,31 +1809,30 @@ LeafNode::erase(context_t c, const search_position_t& pos, bool get_next)
   LOG_PREFIX(OTree::LeafNode::erase);
   assert(!pos.is_end());
   assert(!impl->is_keys_empty());
-  Ref<Node> this_ref = this;
   DEBUGT("erase {}'s pos({}), get_next={} ...",
          c.t, get_name(), pos, get_next);
 
   // get the next cursor
-  return seastar::now().then([c, &pos, get_next, this] {
+  return seastar::now().then([this, c, &pos, get_next] {
     if (get_next) {
       return get_next_cursor(c, pos);
     } else {
       return eagain_ertr::make_ready_future<Ref<tree_cursor_t>>();
     }
-  }).safe_then([c, &pos, this_ref = std::move(this_ref),
-                this, FNAME] (Ref<tree_cursor_t> next_cursor) mutable {
+  }).safe_then([this, c, &pos, FNAME] (Ref<tree_cursor_t> next_cursor) {
```

Comment:
capturing `this` does not help in this case. it's but a raw pointer. it is the refcount which manages lifecycle. that's why we need to capture the `this_ref` by moving it, as we don't want to increase its refcount for no reason -- the callee actually checks the refcount using `assert()` when trying to erase the refcounted node in memory.

Reply:
seems nobody need to check this_ref from the origin place to the new place.  and no others want to erase the leafnode, it is only can be erased by itself, right?  
And I still don't understand why we need this_ref, Yingxin said each transaction has its own copy of tree nodes, that means no tree node is shared between transactions. The node reference is only held because the function call not return.   So why we need check it, jut retire extent will be OK, since no two transactions access the same node.

Reply:
> seems nobody need to check this_ref from the origin place to the new place. and no others want to erase the leafnode, it is only can be erased by itself, right?

i haven't audit the all the related code in fl-tree. but that's my understand per the discussion with Yingxin the other day.

> And I still don't understand why we need this_ref, Yingxin said each transaction has its own copy of tree nodes, that means no tree node is shared between transactions. The node reference is only held because the function call not return. So why we need check it, jut retire extent will be OK, since no two transactions access the same node.

`this_ref` serves two purpose

1. to keep this alive
2. to understand how many strong references it has. so we can erase it without worrying breaking these reference. that's what the failed `assert()` verifies.

Reply:
I need discussion with Yingxin, if we really need check the reference since node is not shared between transactions.

Reply:
Checking `this_ref` counter has nothing to do with transaction sharing, it is a defensive programming that ensures the `Node` is destructed right after it is erased, by ensuring the caller holds the very last reference of the node to earse.

Moving is needed so that the logic doesn't introduce unnecessary increase of reference counters that may break the above assumptions and checks.

Reply:
@tchaikov I have talked with Yingxin, Yingxin agreed this_ref is for his code debugging, when he meets some bugs, he need some assistant method to  locate the issue.  That is not feature's required.  We can keep it as it now.  But I hope we can improve it later.